src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,136 @@
+        }
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> current = getNodeAtIndex(index);
+            Node<T> newNode = new Node<>(current, value, current.prev);
+            if (current.prev == null) {
+                head = newNode;
+            } else {
+                current.prev.next = newNode;
+            }
+            current.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeAtIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> current = getNodeAtIndex(index);
+        T oldValue = current.item;
+        current.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> nodeToRemove = getNodeAtIndex(index);
+        T removedItem = nodeToRemove.item;
+        unlink(nodeToRemove);
+        return removedItem;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        while (current != null) {
+            if ((object == null && current.item == null)
+                    || (object != null && object.equals(current.item))) {
+                unlink(current);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNodeAtIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+        Node<T> current;
+        if (index < size / 2) {
+            current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        } else {
+            current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
+        }
+        return current;
+    }
+
+    private void unlink(Node<T> node) {
+        if (node.prev != null) {
+            node.prev.next = node.next;
+        } else {
+            head = node.next;
+        }
+        if (node.next != null) {
+            node.next.prev = node.prev;
+        } else {
+            tail = node.prev;
+        }
+        size--;
+    }
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        Node(Node<T> next, T item, Node<T> prev) {
+            this.next = next;
+            this.item = item;
+            this.prev = prev;
+        }
     }
```

Comment:
The Node class should be an inner class, not a static nested class. This is because we want to have direct access to the Node's fields without the need for getters and setters, which is possible with an inner class.